airflow/contrib/operators/dataproc_operator.py

```diff
@@ -243,8 +243,92 @@ def __init__(self,
+            projectId=self.project_id,
+            region=self.region
+        ).execute()
+        return result.get('clusters', [])
+
+    def _get_cluster(self, service):
+        cluster_list = self._get_cluster_list_for_project(service)
+        cluster = [c for c in cluster_list if c['clusterName'] == self.cluster_name]
+        if cluster:
+            return cluster[0]
+        return None
+
+    def _get_cluster_state(self, service):
+        cluster = self._get_cluster(service)
+        if cluster and 'status' in cluster:
+            return cluster['status']['state']
+        else:
+            return None
+
+    def _cluster_ready(self, state, service):
+        if state == 'RUNNING':
+            return True
+        if state == 'DELETING':
+            raise Exception('Tried to create a cluster but it\'s in DELETING, something went wrong.')
+        if state == 'ERROR':
+            cluster = self._get_cluster(service)
+            try:
+                error_details = cluster['status']['details']
+            except KeyError:
+                error_details = 'Unknown error in cluster creation, ' \
+                                'check Google Cloud console for details.'
+
+            self.log.info('Dataproc cluster creation resulted in an ERROR state running diagnostics')
+            self.log.info(error_details)
+            diagnose_operation_name = \
+                DataprocClusterCreateOperator._execute_dataproc_diagnose(service, self.project_id,
+                                                                         self.region, self.cluster_name)
+            diagnose_result = DataProcHook._wait_for_operation_done(service, diagnose_operation_name)
+            if diagnose_result.get('response') and diagnose_result.get('response').get('outputUri'):
+                outputUri = diagnose_result.get('response').get('outputUri')
+                self.log.info('Diagnostic information for ERROR cluster available at [%s]', outputUri)
+            else:
+                self.log.info('Diagnostic information could not be retrieved!')
+
+            raise Exception(error_details)
+        return False
+
+    @staticmethod
+    def _execute_dataproc_diagnose(service, project_id, region, cluster_name):
+        response = service.projects().regions().clusters().diagnose(
+            projectId=project_id,
+            region=region,
+            clusterName=cluster_name,
+            body={}
+        ).execute()
+        operation_name = response['name']
+        return operation_name
+
+    def _get_final_cluster_state(self, service):
+        while True:
+            state = self._get_cluster_state(service)
+            if state is None:
+                self.log.info("No state for cluster '%s'", self.cluster_name)
+                time.sleep(15)
+            else:
+                self.log.info("State for cluster '%s' is %s", self.cluster_name, state)
+                return state
+
+    def _wait_for_done(self, service):
```

Comment:
I think this piece of code will be better suited to the hook. In addition, it is worth adding a mechanism to repeat requests
```suggestion
        ).execute(num_retries=self.num_retries)
```